Libraries/LibJS/Runtime/Object.cpp

```diff
@@ -1338,6 +1345,41 @@ void Object::define_intrinsic_accessor(PropertyKey const& property_key, Property
     intrinsics.set(property_key.as_string(), move(accessor));
 }
 
+ThrowCompletionOr<void> Object::for_each_own_property_with_enumerability(Function<ThrowCompletionOr<void>(PropertyKey const&, Value, bool)>&& callback) const
+{
+    auto& vm = this->vm();
+    if (eligible_for_own_property_enumeration_fast_path()) {
+        for (auto& entry : m_indexed_properties) {
+            auto key = PrimitiveString::create(vm, String::number(entry.index()));
+            TRY(callback(entry.index(), key, entry.enumerable()));
+        }
+
+        for (auto const& [property_key, metadata] : shape().property_table()) {
+            if (!property_key.is_string())
+                continue;
+            TRY(callback(property_key, property_key.to_value(vm), metadata.attributes.is_enumerable()));
+        }
```

Comment:
What happens if we add/remove object properties while these loops are running? Do we have test coverage?

Reply:
in `Object::enumerable_own_property_names()` where that does matter fast enumerability check is guarded by `shape().is_cacheable() && &shape() == &pre_iteration_shape`. it's covered by test262. for `for..in` we don't have to bother about that, because `next()` rechecks if property is still defined and enumerable.

Reply:
I added a new test that verifies deleting a key from a getter is reflected in `Object.entries()` result.